packages/next/pages/_document.tsx

```diff
@@ -87,6 +88,59 @@ function hasComponentProps(child: any): child is React.ReactElement {
   return !!child && !!child.props
 }
 
+function handleDocumentScriptLoaderItems(
+  scriptLoader: { beforeInteractive?: any[] },
+  __NEXT_DATA__: NEXT_DATA,
+  props: any
+): void {
+  if (!props.children) return
+
+  const scriptLoaderItems: ScriptProps[] = []
+
+  const children = Array.isArray(props.children)
+    ? props.children
+    : [props.children]
+
+  const headChildren = children.find(
+    (child: React.ReactElement) => child.type === Head
+  )?.props?.children
+  const bodyChildren = children.find(
+    (child: React.ReactElement) => child.type === 'body'
+  )?.props?.children
+
+  // Scripts with beforeInteractive can be placed inside Head or <body> so children of both needs to be traversed
+  const combinedChildren = [
+    ...(Array.isArray(headChildren) ? headChildren : [headChildren]),
+    ...(Array.isArray(bodyChildren) ? bodyChildren : [bodyChildren]),
+  ]
+
+  React.Children.forEach(combinedChildren, (child: any) => {
+    if (!child) return
+
+    if (child.type === Script) {
```

Comment:
Is this intended that only Script tags are allow here?

I had my code to check whether a certain env variables set or not before including Script tag and it didn't work. Here my code sample:

```tsx
class MyDocument extends Document {
    render() {
        return (
            <Html lang="en" className="scroll-smooth">
                <Head>
                    <link rel="stylesheet" href="/main.css" />
                    {process.env.NEXT_PUBLIC_GTAG_ID && (
                        <>
                            <Script
                                strategy="beforeInteractive"
                                src={`https://www.googletagmanager.com/gtag/js?id=${process.env.NEXT_PUBLIC_GTAG_ID}`}
                            />
                            <Script
                                strategy="beforeInteractive"
                                dangerouslySetInnerHTML={{
                                    __html: `
                                        window.dataLayer = window.dataLayer || [];
                                        function gtag(){dataLayer.push(arguments);}
                                        gtag('js', new Date());
                                        gtag('config', '${process.env.NEXT_PUBLIC_GTAG_ID}');
                                    `,
                                }}
                            />
                        </>
                    )}
                </Head>
                <body className="bg-white text-gray-500 antialiased dark:bg-gray-900 dark:text-white">
                    <Main />
                    <NextScript />
                </body>
            </Html>
        );
    }
}
``` 
Because there were two Script tags so I wrapped them with a React.Fragment tag.

**Question:**

 - Can we recursively check all the Script tags here?

Reply:
Thanks for spotting this. Not sure about having it work recursively, but It should work when wrapped in at least one level of `<Fragment>` since that's the [expected behavior for other elements](https://nextjs.org/docs/api-reference/next/head#:~:text=title%2C%20meta%20or%20any%20other%20elements%20(e.g.%20script)%20need%20to%20be%20contained%20as%20direct%20children%20of%20the%20Head%20element%2C%20or%20wrapped%20into%20maximum%20one%20level%20of%20%3CReact.Fragment%3E%20or%20arrays%E2%80%94otherwise%20the%20tags%20won%27t%20be%20correctly%20picked%20up%20on%20client%2Dside%20navigations.) in `<Head>` .

Can you be open a new issue for this? I'll submit a patch to fix, but you can remove the fragment as a workaround for now (sorry for the inconvenience):

```jsx
<Head>
  <link rel="stylesheet" href="/main.css" />
  {process.env.NEXT_PUBLIC_GTAG_ID && (
    <Script
      strategy="beforeInteractive"
      src={`https://www.googletagmanager.com/gtag/js?id=${process.env.NEXT_PUBLIC_GTAG_ID}`}
    />
  )}
  {process.env.NEXT_PUBLIC_GTAG_ID && (
    <Script
      strategy="beforeInteractive"
      dangerouslySetInnerHTML={{
        __html: `
            window.dataLayer = window.dataLayer || [];
            function gtag(){dataLayer.push(arguments);}
            gtag('js', new Date());
            gtag('config', '${process.env.NEXT_PUBLIC_GTAG_ID}');
        `,
      }}
    />
  )}
</Head>;
```

Reply:
#38242